### PR TITLE
[Wait for #2209][WIP][POC] Implement picoGPT/GPT2 Encoder @open sesame 10/10 18:42

### PR DESCRIPTION
**Pull request by DonghakPark**
**Implementing picoGPT/GPT2's Encoder in CPP**
- Using nlohman/json.hpp file so we need to add or make some path to compile json parser
- Need to check utf-8 encoding result

### CODE REVIEW

**Comment by myungjoo** [comment]
It appears that you are direct importing files from other repositories, right?

**Comment by DonghakPark** [comment]
> It appears that you are direct importing files from other repositories, right?

Yes. is it better to make some script files to prepare this kinds of files?

**Comment by myungjoo** [comment]
Great! We are going to personalize GPT on devices!

**Review by jijoongmoon — CHANGES_REQUESTED** [review]
Instead of directly include the external files like json.hpp or ctr-unicode.hpp, we could place them in gihtub.com/nnstreamer/nnstreamer-android-resource/raw/master/external/ repository and download them using shell script like jni/prepare_openblas.sh.

File: meson.build [hunk 1]
```diff
@@ -172,6 +172,13 @@ if get_option('enable-cublas')
   extra_defines += '-DUSE_CUBLAS=1'
 endif
 
+if get_option('platform') != 'tizen'
```
Comment (jijoongmoon) [inline meson.build:175]: I don't think we always use encoder. It only depends on the application Pico. So please remove the running shell script here and add it to the applications.

File: meson_options.txt [hunk 2]
```diff
@@ -33,6 +33,7 @@ option('reduce-tolerance', type: 'boolean', value: true)
 option('enable-long-test', type: 'boolean', value: false)
 
 # backend options
+option('enable-encoder', type: 'boolean', value: true)
```
Comment (jijoongmoon) [inline meson_options.txt:36]: We don't need to add enable-encoder here.

File: packaging/nntrainer.spec [hunk 3]
```diff
@@ -12,7 +12,7 @@
 %define         gen_input $(pwd)/test/input_gen/genInput.py
 %define         support_data_augmentation_opencv 1
 %define         configure_subplugin_install_path -Dnnstreamer-subplugin-install-path=%{nnstreamer_subplugin_path}
-
+%define         use_encoder 1
```
Comment (jijoongmoon) [inline packaging/nntrainer.spec:15]: We don't need to define use_encoder here.

File: packaging/nntrainer.spec [hunk 4]
```diff
@@ -347,6 +347,11 @@ NNSteamer tensor trainer static package for nntrainer to support inference.
 %define enable_cblas -Denable-blas=true
 %endif
 
+# enable encoder
+%if 0%{?use_encoder}
```
Comment (jijoongmoon) [inline packaging/nntrainer.spec:351]: This encoder depends on Applicaiton PicoGPT.. so.. please run the script when we build Applicatin PicoGPT

File: meson.build (hunk 1 above)
Comment (DonghakPark) [reply meson.build:175]: Thank you for Review - Fix

File: meson_options.txt (hunk 2 above)
Comment (DonghakPark) [reply meson_options.txt:36]: Thank you for Review - Fix

File: packaging/nntrainer.spec (hunk 3 above)
Comment (DonghakPark) [reply packaging/nntrainer.spec:15]: Thank you for Review - Fix

File: packaging/nntrainer.spec (hunk 4 above)
Comment (DonghakPark) [reply packaging/nntrainer.spec:351]: Thank you for Review - Fix

**Review by jijoongmoon — APPROVED** [review]
LGTM!

**Review by baek2sm — APPROVED** [review]
LGTM

**Review by Hyoungsukim — APPROVED** [review]
lgtm

**Review by djeong20 — APPROVED** [review]
LGTM

**Pull request merged**